exp/runtime/internal/controllers/warmup_test.go

```diff
@@ -0,0 +1,151 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	utilfeature "k8s.io/component-base/featuregate/testing"
+
+	runtimev1 "sigs.k8s.io/cluster-api/exp/runtime/api/v1alpha1"
+	runtimehooksv1 "sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1"
+	"sigs.k8s.io/cluster-api/feature"
+	"sigs.k8s.io/cluster-api/internal/runtime/catalog"
+	runtimeclient "sigs.k8s.io/cluster-api/internal/runtime/client"
+	runtimev1registry "sigs.k8s.io/cluster-api/internal/runtime/registry"
+)
+
+func Test_warmupRunnable_Start(t *testing.T) {
+	g := NewWithT(t)
+	defer utilfeature.SetFeatureGateDuringTest(t, feature.Gates, feature.ClusterTopology, true)()
+	defer utilfeature.SetFeatureGateDuringTest(t, feature.Gates, feature.RuntimeSDK, true)()
+
+	t.Run("succeed to warm up registry on Start", func(t *testing.T) {
+		ns, err := env.CreateNamespace(ctx, "test-runtime-extension")
+		g.Expect(err).ToNot(HaveOccurred())
+
+		cat := catalog.New()
+		registry := runtimev1registry.New()
+		g.Expect(runtimehooksv1.AddToCatalog(cat)).To(Succeed())
+
+		for _, name := range []string{"ext1", "ext2", "ext3"} {
+			server := fakeExtensionServer(discoveryHandler("first", "second", "third"))
+			defer server.Close()
+			g.Expect(env.CreateAndWait(ctx, fakeExtensionConfigForURL(ns.Name, name, server.URL))).To(Succeed())
+
+			defer func(namespace, name, url string) {
+				_ = env.CleanupAndWait(ctx, fakeExtensionConfigForURL(namespace, name, url))
+			}(ns.Name, name, server.URL)
+		}
+
+		r := &warmupRunnable{
+			Client:    env.GetClient(),
+			APIReader: env.GetAPIReader(),
+			RuntimeClient: runtimeclient.New(runtimeclient.Options{
+				Catalog:  cat,
+				Registry: registry,
+			}),
+		}
+
+		if err := r.Start(ctx); err != nil {
+			t.Error(err)
+		}
+		list := &runtimev1.ExtensionConfigList{}
+		g.Expect(env.GetAPIReader().List(ctx, list)).To(Succeed())
+		g.Expect(len(list.Items)).To(Equal(3))
+		for i, config := range list.Items {
+			// Expect three handlers for each extension and expect the name to be the handler name plus the extension name.
+			handlers := config.Status.Handlers
+			g.Expect(len(handlers)).To(Equal(3))
+			g.Expect(handlers[0].Name).To(Equal(fmt.Sprintf("first.ext%d", i+1)))
+			g.Expect(handlers[1].Name).To(Equal(fmt.Sprintf("second.ext%d", i+1)))
+			g.Expect(handlers[2].Name).To(Equal(fmt.Sprintf("third.ext%d", i+1)))
+
+			conditions := config.GetConditions()
+			g.Expect(len(conditions)).To(Equal(1))
+			g.Expect(conditions[0].Status).To(Equal(corev1.ConditionTrue))
+			g.Expect(conditions[0].Type).To(Equal(runtimev1.RuntimeExtensionDiscoveredCondition))
+		}
+	})
+
+	t.Run("fail to warm up registry on Start with uncontactable extension", func(t *testing.T) {
```

Comment:
nit: WDYT about `unreachable` or `broken`? (same would apply to l.99+
```suggestion
	t.Run("fail to warm up registry on Start with unreachable extension", func(t *testing.T) {
```